### PR TITLE
gc_worker: add integration tests to lock observer

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?
Problem Summary:
Existing tests don't cover the raftstore. And `test_collect_lock_from_stale_leader` covers https://github.com/tikv/tikv/issues/8184.

### What is changed and how it works?
What's Changed:
Add integration tests to lock observer which cover applying logs and applying snapshot.

### Related changes
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 
* No release note

### CODE REVIEW

**Comment by youjiali1995** [comment]
@MyonKeminta PTAL.

File: tests/failpoints/cases/test_gc_worker.rs [hunk 1]
```diff
@@ -74,3 +77,184 @@ fn test_gcworker_busy() {
...
+        &client,
+        ctx.clone(),
+        vec![new_mutation(Op::Put, b"k2", b"v")],
+        b"k1".to_vec(),
+        10,
+    );
+    let resp = check_lock_observer(&client, max_ts);
+    assert!(resp.get_error().is_empty(), "{:?}", resp.get_error());
+    // Should mark clean if fails to send locks.
+    assert!(!resp.get_is_clean());
```
Comment (MyonKeminta) [inline tests/failpoints/cases/test_gc_worker.rs:259]: I'm afraid there may be some issue when checking happens between send error occurs and marking dirty.

Comment (youjiali1995) [reply tests/failpoints/cases/test_gc_worker.rs:259]: IMO, it's not an issue.

File: src/server/gc_worker/applied_lock_collector.rs [hunk 2]
```diff
@@ -123,10 +131,21 @@ impl LockObserver {
...
     fn send(&self, locks: Vec<(Key, Lock)>) {
-        match self
+        let res = &mut self
             .sender
-            .schedule(LockCollectorTask::ObservedLocks(locks))
-        {
+            .schedule(LockCollectorTask::ObservedLocks(locks));
+        // Wrapping the fail point in a closure, so we can modify
+        // local variables without return,
+        let mut send_fp = || {
```
Comment (sticnarf) [inline src/server/gc_worker/applied_lock_collector.rs:139]: There's a warning on this line:
```
warning: variable does not need to be mutable
   --> src/server/gc_worker/applied_lock_collector.rs:139:13
    |
139 |         let mut send_fp = || {
    |             ----^^^^^^^
    |             |
    |             help: remove this `mut`
    |
    = note: `#[warn(unused_mut)]` on by default
```

However, CI passed?

Comment (youjiali1995) [reply src/server/gc_worker/applied_lock_collector.rs:139]: Will be fixed by https://github.com/tikv/tikv/pull/8217/files#diff-dd25bab50f5cfdaeeb301092b7820da3R139.

**Pull request merged**